packages/template-retail-react-app/app/pages/checkout/util/checkout-context.js

```diff
@@ -4,98 +4,56 @@
+    const STEPS = CHECKOUT_STEPS_LIST.reduce((acc, step, idx) => ({...acc, [step]: idx}), {})
 
-    const getCheckoutStepName = (step) => CheckoutStepsList[step]
-
-    const mergeState = useCallback((data) => {
-        // If we become unmounted during an async call that results in updating state, we
-        // skip the update to avoid React errors about setting state in unmounted components.
-        if (!mounted.current) {
-            return
-        }
-        setState((_state) => ({
-            ..._state,
-            ...data
-        }))
-    })
-
-    // We use this to track mounted state.
-    useEffect(() => {
-        mounted.current = true
-        return () => {
-            mounted.current = false
-        }
-    }, [])
-
-    useEffect(() => {
-        if (customer.isRegistered && state.isGuestCheckout) {
-            mergeState({isGuestCheckout: false})
-        }
-
-        if (customer.isGuest && basket?.customerInfo?.email && !state.isGuestCheckout) {
-            mergeState({isGuestCheckout: true})
-        }
+    const getCheckoutStepName = (step) => CHECKOUT_STEPS_LIST[step]
 
-        // Derive the starting step for checkout based on current state of basket.
-        // A failed condition sets the current step and returns early (order matters).
-        if (customer.customerId && basket?.basketId && state.step === undefined) {
-            if (!basket.customerInfo?.email) {
-                mergeState({step: CheckoutSteps.Contact_Info})
+    useEffect(
+        () => {
+            if (!customer || !basket) {
                 return
             }
-            if (basket.shipments && !basket.shipments[0]?.shippingAddress) {
-                mergeState({step: CheckoutSteps.Shipping_Address})
-                return
-            }
-            if (basket.shipments && !basket.shipments[0]?.shippingMethod) {
-                mergeState({step: CheckoutSteps.Shipping_Options})
-                return
-            }
-            if (!basket.paymentInstruments || !basket.billingAddress) {
-                mergeState({step: CheckoutSteps.Payment})
-                return
+
+            let step = STEPS.REVIEW_ORDER
+
+            if (customer.isGuest && !basket.customerInfo?.email) {
+                step = STEPS.CONTACT_INFO
+            } else if (!basket.shipments[0]?.shippingAddress) {
+                step = STEPS.SHIPPING_ADDRESS
+            } else if (!basket.shipments[0]?.shippingMethod) {
+                step = STEPS.SHIPPING_OPTIONS
+            } else if (!basket.paymentInstruments || !basket.billingAddress) {
+                step = STEPS.PAYMENT
             }
 
-            mergeState({step: CheckoutSteps.Review_Order})
-        }
-    }, [customer, basket])
+            setStep(step)
+        },
+        [customer?.isGuest, basket?.customerInfo?.email, basket?.shipments[0]?.shippingAddress],
+        basket?.shipments[0]?.shippingMethod,
+        basket?.paymentInstruments,
+        basket?.billingAddress
```

Comment:
What are these additional parameters to the `useEffect` for? Usually it's just 1 or 2 parameters.

Reply:
these useEffect dependencies are used inside the useEffect, if those state changes, it triggers the useEffect. I believe react does a `===` comparison and it doesn't work well for objects. This is to avoid bugs since I recently saw a [bug](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/1061/files#diff-437f4419ee9ce02d121bad9b7a635759b9939c3f75412a88cb7feae15a991a45L74-L82) that causes infinite re-rendering.

Reply:
If those are also dependencies, that they need to go as the 2nd parameter to useEffect, right?

What I meant earlier is currently we have 5 parameters to `useEffect`. So I think what you want is to combine all those dependencies into the array.

Reply:
These are the 2nd parameter and in an array. Not sure if i understand you correctly?

Reply:
My bad! my eyes are not working